File: backend/src/services/digikey.rs
use anyhow::{Context, Result};
use once_cell::sync::Lazy;
use reqwest::Client;
use serde::{Deserialize, Serialize};
use std::sync::RwLock;
use std::time::{Duration, Instant};
use tracing::{debug, error, info, warn};

use crate::types::{DigiKeyParameter, DigiKeyPartInfo};

// DigiKey API Configuration
static DIGIKEY_CLIENT_ID: Lazy<String> = Lazy::new(|| {
    std::env::var("DIGIKEY_CLIENT_ID").unwrap_or_else(|_| {
        warn!("DIGIKEY_CLIENT_ID not set - DigiKey integration will not work");
        String::new()
    })
});

static DIGIKEY_CLIENT_SECRET: Lazy<String> = Lazy::new(|| {
    std::env::var("DIGIKEY_CLIENT_SECRET").unwrap_or_else(|_| {
        warn!("DIGIKEY_CLIENT_SECRET not set - DigiKey integration will not work");
        String::new()
    })
});

// DigiKey API endpoints
const DIGIKEY_AUTH_URL: &str = "https://api.digikey.com/v1/oauth2/token";
const DIGIKEY_SEARCH_URL: &str = "https://api.digikey.com/products/v4/search/keyword";

// Token cache with thread-safe access
static TOKEN_CACHE: Lazy<RwLock<Option<TokenCache>>> = Lazy::new(|| RwLock::new(None));

// Shared HTTP client - reqwest Client uses connection pooling internally
static HTTP_CLIENT: Lazy<Client> = Lazy::new(|| {
    Client::builder()
        .timeout(Duration::from_secs(30))
        .build()
        .expect("Failed to create HTTP client")
});

#[derive(Debug, Clone)]
struct TokenCache {
    access_token: String,
    expires_at: Instant,
}

#[derive(Debug, Deserialize)]
struct TokenResponse {
    access_token: String,
    expires_in: u64,
    #[allow(dead_code)]
    token_type: String,
}

#[derive(Debug, Serialize)]
struct KeywordSearchRequest {
    #[serde(rename = "Keywords")]
    keywords: String,
    #[serde(rename = "RecordCount")]
    record_count: i32,
    #[serde(rename = "RecordStartPosition")]
    record_start_position: i32,
}

#[derive(Debug, Deserialize)]
struct DigiKeySearchResponse {
    #[serde(rename = "Products")]
    products: Option<Vec<DigiKeyProduct>>,
    #[serde(rename = "ProductsCount")]
    #[allow(dead_code)]
    products_count: Option<i32>,
    #[serde(rename = "ExactManufacturerProducts")]
    exact_manufacturer_products: Option<Vec<DigiKeyProduct>>,
    #[serde(rename = "ExactManufacturerProductsCount")]
    #[allow(dead_code)]
    exact_manufacturer_products_count: Option<i32>,
    #[serde(rename = "ExactDigiKeyProduct")]
    exact_digikey_product: Option<DigiKeyProduct>,
}

#[derive(Debug, Deserialize)]
struct DigiKeyProduct {
    // Note: v4 API doesn't include DigiKey part number in the product object
    // It's available through ProductVariations
    #[serde(rename = "ManufacturerProductNumber")]
    manufacturer_part_number: Option<String>,
    #[serde(rename = "Manufacturer")]
    manufacturer: Option<ManufacturerInfo>,
    // Description is an object containing ProductDescription and DetailedDescription
    #[serde(rename = "Description")]
    description: Option<DescriptionInfo>,
    #[serde(rename = "ProductUrl")]
    product_url: Option<String>,
    #[serde(rename = "DatasheetUrl")]
    datasheet_url: Option<String>,
    #[serde(rename = "PhotoUrl")]
    primary_photo: Option<String>,
    #[serde(rename = "QuantityAvailable")]
    quantity_available: Option<i64>,
    #[serde(rename = "UnitPrice")]
    unit_price: Option<f64>,
    #[serde(rename = "ProductStatus")]
    product_status: Option<ProductStatus>,
    #[serde(rename = "Category")]
    category: Option<CategoryInfo>,
    #[serde(rename = "Parameters")]
    parameters: Option<Vec<ParameterInfo>>,
    // v4 has ProductVariations which contains the DigiKey part numbers
    #[serde(rename = "ProductVariations")]
    product_variations: Option<Vec<ProductVariation>>,
}

#[derive(Debug, Deserialize)]
struct DescriptionInfo {
    #[serde(rename = "ProductDescription")]
    product_description: Option<String>,
    #[serde(rename = "DetailedDescription")]
    detailed_description: Option<String>,
}

#[derive(Debug, Deserialize)]
struct ProductVariation {
    #[serde(rename = "DigiKeyProductNumber")]
    digikey_product_number: Option<String>,
    #[serde(rename = "PackageType")]
    #[allow(dead_code)]
    package_type: Option<PackageType>,
}

#[derive(Debug, Deserialize)]
#[allow(dead_code)]
struct PackageType {
    #[serde(rename = "Name")]
    name: Option<String>,
}

#[derive(Debug, Deserialize)]
struct ManufacturerInfo {
    #[serde(rename = "Name")]
    name: Option<String>,
}

#[derive(Debug, Deserialize)]
struct ProductStatus {
    #[serde(rename = "Status")]
    status: Option<String>,
    #[serde(rename = "Id")]
    #[allow(dead_code)]
    id: Option<i32>,
}

#[derive(Debug, Deserialize)]
struct CategoryInfo {
    #[serde(rename = "Name")]
    name: Option<String>,
}

#[derive(Debug, Deserialize)]
struct ParameterInfo {
    #[serde(rename = "ParameterText")]
    parameter_text: Option<String>,
    #[serde(rename = "ValueText")]
    value_text: Option<String>,
}

pub struct DigiKeyClient;

impl DigiKeyClient {
    pub fn new() -> Self {
        Self
    }

    /// Check if DigiKey integration is configured
    pub fn is_configured() -> bool {
        !DIGIKEY_CLIENT_ID.is_empty() && !DIGIKEY_CLIENT_SECRET.is_empty()
    }

    /// Get a valid access token, refreshing if necessary
    async fn get_access_token(&self) -> Result<String> {
        // Check cache first
        {
            let cache = TOKEN_CACHE.read().unwrap();
            if let Some(ref cached) = *cache {
                // Use token if it has at least 60 seconds remaining
                if cached.expires_at > Instant::now() + Duration::from_secs(60) {
                    debug!("Using cached DigiKey access token");
                    return Ok(cached.access_token.clone());
                }
            }
        }

        // Need to refresh token
        info!("Refreshing DigiKey access token");
        
        let params = [
            ("client_id", DIGIKEY_CLIENT_ID.as_str()),
            ("client_secret", DIGIKEY_CLIENT_SECRET.as_str()),
            ("grant_type", "client_credentials"),
        ];

        let response = HTTP_CLIENT
            .post(DIGIKEY_AUTH_URL)
            .form(&params)
            .send()
            .await
            .context("Failed to send token request to DigiKey")?;

        if !response.status().is_success() {
            let status = response.status();
            let body = response.text().await.unwrap_or_default();
            error!("DigiKey auth failed: {} - {}", status, body);
            anyhow::bail!("DigiKey authentication failed: {} - {}", status, body);
        }

        let token_response: TokenResponse = response
            .json()
            .await
            .context("Failed to parse DigiKey token response")?;

        let expires_at = Instant::now() + Duration::from_secs(token_response.expires_in);
        
        // Update cache
        {
            let mut cache = TOKEN_CACHE.write().unwrap();
            *cache = Some(TokenCache {
                access_token: token_response.access_token.clone(),
                expires_at,
            });
        }

        info!("DigiKey access token refreshed successfully");
        Ok(token_response.access_token)
    }

    /// Search for parts by keyword/MPN
    /// This uses the keyword search endpoint which is more flexible and returns
    /// exact manufacturer matches when searching by MPN
    pub async fn search_keyword(&self, query: &str) -> Result<Vec<DigiKeyPartInfo>> {
        if !Self::is_configured() {
            anyhow::bail!("DigiKey API not configured. Set DIGIKEY_CLIENT_ID and DIGIKEY_CLIENT_SECRET environment variables.");
        }

        let access_token = self.get_access_token().await?;

        let request_body = KeywordSearchRequest {
            keywords: query.to_string(),
            record_count: 10,
            record_start_position: 0,
        };

        debug!("Searching DigiKey for: {}", query);

        let response = HTTP_CLIENT
            .post(DIGIKEY_SEARCH_URL)
            .header("Authorization", format!("Bearer {}", access_token))
            .header("X-DIGIKEY-Client-Id", DIGIKEY_CLIENT_ID.as_str())
            .header("X-DIGIKEY-Locale-Site", "US")
            .header("X-DIGIKEY-Locale-Language", "en")
            .header("X-DIGIKEY-Locale-Currency", "USD")
            .json(&request_body)
            .send()
            .await
            .context("Failed to send search request to DigiKey")?;

        if !response.status().is_success() {
            let status = response.status();
            let body = response.text().await.unwrap_or_default();
            error!("DigiKey search failed: {} - {}", status, body);
            anyhow::bail!("DigiKey search failed: {} - {}", status, body);
        }

        let search_response: DigiKeySearchResponse = response
            .json()
            .await
            .context("Failed to parse DigiKey search response")?;

        // Prioritize exact manufacturer matches over general keyword matches
        // This gives better results when searching by MPN (e.g., "ESP32-WROOM-32E-N4")
        let mut parts = Vec::new();

        // First, add exact DigiKey product if found
        if let Some(exact_dk) = search_response.exact_digikey_product {
            info!("Found exact DigiKey product match");
            parts.push(Self::convert_product(exact_dk));
        }

        // Then add exact manufacturer matches (best for MPN searches)
        if let Some(exact_mfr) = search_response.exact_manufacturer_products {
            let count = exact_mfr.len();
            info!("Found {} exact manufacturer product matches", count);
            for product in exact_mfr {
                parts.push(Self::convert_product(product));
            }
        }

        // Finally add general keyword matches if we don't have enough results
        if parts.len() < 5 {
            if let Some(products) = search_response.products {
                let remaining = 10 - parts.len();
                for product in products.into_iter().take(remaining) {
                    // Avoid duplicates by checking part numbers
                    let mpn = product.manufacturer_part_number.as_ref();
                    let already_included = parts.iter().any(|p| {
                        p.manufacturer_part_number.as_ref() == mpn
                    });
                    if !already_included {
                        parts.push(Self::convert_product(product));
                    }
                }
            }
        }

        info!("DigiKey search returned {} total parts", parts.len());
        Ok(parts)
    }

    /// Convert DigiKey API product to our internal representation
    fn convert_product(product: DigiKeyProduct) -> DigiKeyPartInfo {
        // Get DigiKey part number from first product variation
        let digikey_part_number = product
            .product_variations
            .as_ref()
            .and_then(|variations| variations.first())
            .and_then(|v| v.digikey_product_number.clone());
        
        debug!(
            "Converting product: DK#={:?}, MPN={:?}, Mfr={:?}",
            digikey_part_number,
            product.manufacturer_part_number,
            product.manufacturer.as_ref().and_then(|m| m.name.as_ref())
        );
        
        let status = product.product_status.as_ref().and_then(|s| s.status.clone());
        let is_obsolete = status
            .as_ref()
            .map(|s| {
                let s_lower = s.to_lowercase();
                s_lower.contains("obsolete")
                    || s_lower.contains("discontinued")
                    || s_lower.contains("not for new designs")
                    || s_lower.contains("last time buy")
            })
            .unwrap_or(false);

        let lifecycle_status = if is_obsolete {
            Some(status.clone().unwrap_or_else(|| "Obsolete".to_string()))
        } else {
            status.clone()
        };

        // Extract descriptions from the Description object
        let (description, detailed_description) = match product.description {
            Some(desc) => (desc.product_description, desc.detailed_description),
            None => (None, None),
        };

        DigiKeyPartInfo {
            digikey_part_number,
            manufacturer_part_number: product.manufacturer_part_number,
            manufacturer: product.manufacturer.and_then(|m| m.name),
            description,
            detailed_description,
            product_url: product.product_url,
            datasheet_url: product.datasheet_url,
            photo_url: product.primary_photo,
            quantity_available: product.quantity_available,
            unit_price: product.unit_price,
            product_status: status,
            is_obsolete,
            lifecycle_status,
            category: product.category.and_then(|c| c.name),
            parameters: product
                .parameters
                .unwrap_or_default()
                .into_iter()
                .filter_map(|p| {
                    Some(DigiKeyParameter {
                        name: p.parameter_text?,
                        value: p.value_text.unwrap_or_default(),
                    })
                })
                .collect(),
        }
    }
}

impl Default for DigiKeyClient {
    fn default() -> Self {
        Self::new()
    }
}

File: kicanvas/worker/digikey.ts
/**
 * DigiKey OAuth 3-Legged Flow Cloudflare Worker
 *
 * This worker implements the DigiKey OAuth 2.0 3-legged authorization flow,
 * allowing users to authenticate with their own DigiKey accounts to query
 * the DigiKey API directly.
 *
 * Flow:
 * 1. User clicks "Connect DigiKey" → redirected to /auth/digikey/login
 * 2. Worker redirects to DigiKey authorization page
 * 3. User authorizes → DigiKey redirects to /auth/digikey/callback
 * 4. Worker exchanges code for tokens, stores in KV, sets session cookie
 * 5. Frontend can now call /api/digikey/* endpoints using session cookie
 *
 * Required environment bindings:
 * - DIGIKEY_CLIENT_ID: DigiKey OAuth client ID
 * - DIGIKEY_CLIENT_SECRET: DigiKey OAuth client secret
 * - DIGIKEY_SESSIONS: KV namespace for storing user sessions/tokens
 */

export interface Env {
    DIGIKEY_CLIENT_ID: string;
    DIGIKEY_CLIENT_SECRET: string;
    DIGIKEY_SESSIONS: KVNamespace;
}

// DigiKey API endpoints
const DIGIKEY_AUTH_URL = "https://api.digikey.com/v1/oauth2/authorize";
const DIGIKEY_TOKEN_URL = "https://api.digikey.com/v1/oauth2/token";
const DIGIKEY_SEARCH_URL = "https://api.digikey.com/products/v4/search/keyword";

// Cookie/session configuration
const SESSION_COOKIE_NAME = "digikey_session";
const SESSION_TTL_SECONDS = 60 * 60 * 24 * 7; // 7 days

// Types
interface TokenResponse {
    access_token: string;
    refresh_token: string;
    expires_in: number;
    token_type: string;
}

interface SessionData {
    access_token: string;
    refresh_token: string;
    expires_at: number; // Unix timestamp
}

interface KeywordSearchRequest {
    Keywords: string;
    RecordCount: number;
    RecordStartPosition: number;
}

interface DigiKeyProduct {
    ManufacturerProductNumber?: string;
    Manufacturer?: { Name?: string };
    Description?: {
        ProductDescription?: string;
        DetailedDescription?: string;
    };
    ProductUrl?: string;
    DatasheetUrl?: string;
    PhotoUrl?: string;
    QuantityAvailable?: number;
    UnitPrice?: number;
    ProductStatus?: { Status?: string; Id?: number };
    Category?: { Name?: string };
    Parameters?: Array<{ ParameterText?: string; ValueText?: string }>;
    ProductVariations?: Array<{
        DigiKeyProductNumber?: string;
        PackageType?: { Name?: string };
    }>;
}

interface DigiKeySearchResponse {
    Products?: DigiKeyProduct[];
    ProductsCount?: number;
    ExactManufacturerProducts?: DigiKeyProduct[];
    ExactManufacturerProductsCount?: number;
    ExactDigiKeyProduct?: DigiKeyProduct;
}

// Utility functions
function generateSessionId(): string {
    const bytes = new Uint8Array(32);
    crypto.getRandomValues(bytes);
    return Array.from(bytes)
        .map((b) => b.toString(16).padStart(2, "0"))
        .join("");
}

function getSessionIdFromRequest(request: Request): string | null {
    const cookieHeader = request.headers.get("Cookie");
    if (!cookieHeader) return null;

    const cookies = cookieHeader.split(";").map((c) => c.trim());
    for (const cookie of cookies) {
        const [name, value] = cookie.split("=");
        if (name === SESSION_COOKIE_NAME) {
            return value;
        }
    }
    return null;
}

function createSessionCookie(
    sessionId: string,
    maxAge: number = SESSION_TTL_SECONDS,
): string {
    return `${SESSION_COOKIE_NAME}=${sessionId}; Path=/; HttpOnly; Secure; SameSite=Lax; Max-Age=${maxAge}`;
}

function corsHeaders(origin: string | null): HeadersInit {
    return {
        "Access-Control-Allow-Origin": origin || "*",
        "Access-Control-Allow-Methods": "GET, POST, OPTIONS",
        "Access-Control-Allow-Headers": "Content-Type",
        "Access-Control-Allow-Credentials": "true",
    };
}

function jsonResponse(
    data: unknown,
    status: number = 200,
    headers: HeadersInit = {},
): Response {
    return new Response(JSON.stringify(data), {
        status,
        headers: {
            "Content-Type": "application/json",
            ...headers,
        },
    });
}

// Convert DigiKey product to our normalized format
function convertProduct(product: DigiKeyProduct) {
    const digikeyPartNumber =
        product.ProductVariations?.[0]?.DigiKeyProductNumber ?? null;

    const status = product.ProductStatus?.Status ?? null;
    const isObsolete = status
        ? /obsolete|discontinued|not for new designs|last time buy/i.test(
              status,
          )
        : false;

    return {
        digikey_part_number: digikeyPartNumber,
        manufacturer_part_number: product.ManufacturerProductNumber ?? null,
        manufacturer: product.Manufacturer?.Name ?? null,
        description: product.Description?.ProductDescription ?? null,
        detailed_description: product.Description?.DetailedDescription ?? null,
        product_url: product.ProductUrl ?? null,
        datasheet_url: product.DatasheetUrl ?? null,
        photo_url: product.PhotoUrl ?? null,
        quantity_available: product.QuantityAvailable ?? null,
        unit_price: product.UnitPrice ?? null,
        product_status: status,
        is_obsolete: isObsolete,
        lifecycle_status: isObsolete ? status ?? "Obsolete" : status,
        category: product.Category?.Name ?? null,
        parameters: (product.Parameters ?? [])
            .filter((p) => p.ParameterText)
            .map((p) => ({
                name: p.ParameterText!,
                value: p.ValueText ?? "",
            })),
    };
}

// Session management
async function getSession(
    env: Env,
    sessionId: string,
): Promise<SessionData | null> {
    const data = await env.DIGIKEY_SESSIONS.get(sessionId, "json");
    return data as SessionData | null;
}

async function saveSession(
    env: Env,
    sessionId: string,
    session: SessionData,
): Promise<void> {
    await env.DIGIKEY_SESSIONS.put(sessionId, JSON.stringify(session), {
        expirationTtl: SESSION_TTL_SECONDS,
    });
}

async function deleteSession(env: Env, sessionId: string): Promise<void> {
    await env.DIGIKEY_SESSIONS.delete(sessionId);
}

// Token management
async function exchangeCodeForTokens(
    env: Env,
    code: string,
    redirectUri: string,
): Promise<TokenResponse> {
    const params = new URLSearchParams({
        code,
        client_id: env.DIGIKEY_CLIENT_ID,
        client_secret: env.DIGIKEY_CLIENT_SECRET,
        redirect_uri: redirectUri,
        grant_type: "authorization_code",
    });

    const response = await fetch(DIGIKEY_TOKEN_URL, {
        method: "POST",
        headers: {
            "Content-Type": "application/x-www-form-urlencoded",
        },
        body: params.toString(),
    });

    if (!response.ok) {
        const error = await response.text();
        throw new Error(`Token exchange failed: ${response.status} - ${error}`);
    }

    return response.json();
}

async function refreshAccessToken(
    env: Env,
    refreshToken: string,
): Promise<TokenResponse> {
    const params = new URLSearchParams({
        refresh_token: refreshToken,
        client_id: env.DIGIKEY_CLIENT_ID,
        client_secret: env.DIGIKEY_CLIENT_SECRET,
        grant_type: "refresh_token",
    });

    const response = await fetch(DIGIKEY_TOKEN_URL, {
        method: "POST",
        headers: {
            "Content-Type": "application/x-www-form-urlencoded",
        },
        body: params.toString(),
    });

    if (!response.ok) {
        const error = await response.text();
        throw new Error(`Token refresh failed: ${response.status} - ${error}`);
    }

    return response.json();
}

// Get a valid access token, refreshing if necessary
async function getValidAccessToken(
    env: Env,
    sessionId: string,
    session: SessionData,
): Promise<string | null> {
    const now = Date.now();
    const bufferMs = 60 * 1000; // 1 minute buffer

    // Token is still valid
    if (session.expires_at > now + bufferMs) {
        return session.access_token;
    }

    // Need to refresh
    try {
        const tokens = await refreshAccessToken(env, session.refresh_token);
        const newSession: SessionData = {
            access_token: tokens.access_token,
            refresh_token: tokens.refresh_token,
            expires_at: now + tokens.expires_in * 1000,
        };
        await saveSession(env, sessionId, newSession);
        return tokens.access_token;
    } catch (error) {
        console.error("Failed to refresh token:", error);
        // Delete invalid session
        await deleteSession(env, sessionId);
        return null;
    }
}

// Route handlers
async function handleLogin(request: Request, env: Env): Promise<Response> {
    const url = new URL(request.url);
    const returnUrl =
        url.searchParams.get("return_url") || url.origin + "/debug/";
    const redirectUri = `${url.origin}/auth/digikey/callback`;

    // Store return URL in state parameter (base64 encoded)
    const state = btoa(JSON.stringify({ return_url: returnUrl }));

    const authUrl = new URL(DIGIKEY_AUTH_URL);
    authUrl.searchParams.set("response_type", "code");
    authUrl.searchParams.set("client_id", env.DIGIKEY_CLIENT_ID);
    authUrl.searchParams.set("redirect_uri", redirectUri);
    authUrl.searchParams.set("state", state);

    return Response.redirect(authUrl.toString(), 302);
}

async function handleCallback(request: Request, env: Env): Promise<Response> {
    const url = new URL(request.url);
    const code = url.searchParams.get("code");
    const state = url.searchParams.get("state");
    const error = url.searchParams.get("error");
    const errorDescription = url.searchParams.get("error_description");

    // Default return URL
    let returnUrl = url.origin + "/debug/";

    // Parse state to get return URL
    if (state) {
        try {
            const stateData = JSON.parse(atob(state));
            if (stateData.return_url) {
                returnUrl = stateData.return_url;
            }
        } catch {
            // Ignore parse errors, use default
        }
    }

    // Handle OAuth errors
    if (error) {
        const redirectUrl = new URL(returnUrl);
        redirectUrl.searchParams.set("digikey_error", error);
        if (errorDescription) {
            redirectUrl.searchParams.set(
                "digikey_error_description",
                errorDescription,
            );
        }
        return Response.redirect(redirectUrl.toString(), 302);
    }

    if (!code) {
        const redirectUrl = new URL(returnUrl);
        redirectUrl.searchParams.set(
            "digikey_error",
            "missing_code",
        );
        return Response.redirect(redirectUrl.toString(), 302);
    }

    try {
        // Exchange code for tokens
        const redirectUri = `${url.origin}/auth/digikey/callback`;
        const tokens = await exchangeCodeForTokens(env, code, redirectUri);

        // Create session
        const sessionId = generateSessionId();
        const session: SessionData = {
            access_token: tokens.access_token,
            refresh_token: tokens.refresh_token,
            expires_at: Date.now() + tokens.expires_in * 1000,
        };
        await saveSession(env, sessionId, session);

        // Redirect back with session cookie
        const redirectUrl = new URL(returnUrl);
        redirectUrl.searchParams.set("digikey_connected", "true");

        const response = Response.redirect(redirectUrl.toString(), 302);
        return new Response(response.body, {
            status: response.status,
            headers: {
                ...Object.fromEntries(response.headers.entries()),
                "Set-Cookie": createSessionCookie(sessionId),
            },
        });
    } catch (err) {
        console.error("OAuth callback error:", err);
        const redirectUrl = new URL(returnUrl);
        redirectUrl.searchParams.set("digikey_error", "token_exchange_failed");
        redirectUrl.searchParams.set(
            "digikey_error_description",
            err instanceof Error ? err.message : "Unknown error",
        );
        return Response.redirect(redirectUrl.toString(), 302);
    }
}

async function handleLogout(request: Request, env: Env): Promise<Response> {
    const origin = request.headers.get("Origin");
    const sessionId = getSessionIdFromRequest(request);

    if (sessionId) {
        await deleteSession(env, sessionId);
    }

    return new Response(JSON.stringify({ success: true }), {
        status: 200,
        headers: {
            "Content-Type": "application/json",
            "Set-Cookie": createSessionCookie("", 0), // Clear cookie
            ...corsHeaders(origin),
        },
    });
}

async function handleStatus(request: Request, env: Env): Promise<Response> {
    const origin = request.headers.get("Origin");
    const sessionId = getSessionIdFromRequest(request);

    if (!sessionId) {
        return jsonResponse(
            {
                connected: false,
                message: "Not connected to DigiKey",
            },
            200,
            corsHeaders(origin),
        );
    }

    const session = await getSession(env, sessionId);
    if (!session) {
        return jsonResponse(
            {
                connected: false,
                message: "Session expired",
            },
            200,
            {
                "Set-Cookie": createSessionCookie("", 0), // Clear invalid cookie
                ...corsHeaders(origin),
            },
        );
    }

    // Check if token is still valid (or can be refreshed)
    const accessToken = await getValidAccessToken(env, sessionId, session);

    return jsonResponse(
        {
            connected: !!accessToken,
            message: accessToken
                ? "Connected to DigiKey"
                : "Session expired, please reconnect",
        },
        200,
        corsHeaders(origin),
    );
}

async function handleSearch(request: Request, env: Env): Promise<Response> {
    const origin = request.headers.get("Origin");
    const sessionId = getSessionIdFromRequest(request);

    if (!sessionId) {
        return jsonResponse(
            {
                success: false,
                error: "Not authenticated. Please connect your DigiKey account.",
                parts: [],
                total_count: 0,
            },
            401,
            corsHeaders(origin),
        );
    }

    const session = await getSession(env, sessionId);
    if (!session) {
        return jsonResponse(
            {
                success: false,
                error: "Session expired. Please reconnect your DigiKey account.",
                parts: [],
                total_count: 0,
            },
            401,
            {
                "Set-Cookie": createSessionCookie("", 0),
                ...corsHeaders(origin),
            },
        );
    }

    const accessToken = await getValidAccessToken(env, sessionId, session);
    if (!accessToken) {
        return jsonResponse(
            {
                success: false,
                error: "Session expired. Please reconnect your DigiKey account.",
                parts: [],
                total_count: 0,
            },
            401,
            {
                "Set-Cookie": createSessionCookie("", 0),
                ...corsHeaders(origin),
            },
        );
    }

    // Parse request body
    let query: string;
    try {
        const body = await request.json();
        query = body.query || body.mpn || "";
        if (!query) {
            return jsonResponse(
                {
                    query: "",
                    success: false,
                    error: "Missing search query",
                    parts: [],
                    total_count: 0,
                },
                400,
                corsHeaders(origin),
            );
        }
    } catch {
        return jsonResponse(
            {
                query: "",
                success: false,
                error: "Invalid request body",
                parts: [],
                total_count: 0,
            },
            400,
            corsHeaders(origin),
        );
    }

    // Make DigiKey API request
    const searchRequest: KeywordSearchRequest = {
        Keywords: query,
        RecordCount: 10,
        RecordStartPosition: 0,
    };

    try {
        const response = await fetch(DIGIKEY_SEARCH_URL, {
            method: "POST",
            headers: {
                Authorization: `Bearer ${accessToken}`,
                "X-DIGIKEY-Client-Id": env.DIGIKEY_CLIENT_ID,
                "X-DIGIKEY-Locale-Site": "US",
                "X-DIGIKEY-Locale-Language": "en",
                "X-DIGIKEY-Locale-Currency": "USD",
                "Content-Type": "application/json",
            },
            body: JSON.stringify(searchRequest),
        });

        if (!response.ok) {
            const errorText = await response.text();
            console.error("DigiKey search failed:", response.status, errorText);
            return jsonResponse(
                {
                    query,
                    success: false,
                    error: `DigiKey API error: ${response.status}`,
                    parts: [],
                    total_count: 0,
                },
                502,
                corsHeaders(origin),
            );
        }

        const searchResponse: DigiKeySearchResponse = await response.json();

        // Build results, prioritizing exact matches
        const parts: ReturnType<typeof convertProduct>[] = [];
        const seenMpns = new Set<string>();

        // First, add exact DigiKey product if found
        if (searchResponse.ExactDigiKeyProduct) {
            const converted = convertProduct(searchResponse.ExactDigiKeyProduct);
            if (converted.manufacturer_part_number) {
                seenMpns.add(converted.manufacturer_part_number);
            }
            parts.push(converted);
        }

        // Then add exact manufacturer matches
        if (searchResponse.ExactManufacturerProducts) {
            for (const product of searchResponse.ExactManufacturerProducts) {
                const converted = convertProduct(product);
                if (
                    converted.manufacturer_part_number &&
                    !seenMpns.has(converted.manufacturer_part_number)
                ) {
                    seenMpns.add(converted.manufacturer_part_number);
                    parts.push(converted);
                }
            }
        }

        // Finally add general keyword matches if we don't have enough
        if (parts.length < 5 && searchResponse.Products) {
            for (const product of searchResponse.Products) {
                if (parts.length >= 10) break;
                const converted = convertProduct(product);
                if (
                    converted.manufacturer_part_number &&
                    !seenMpns.has(converted.manufacturer_part_number)
                ) {
                    seenMpns.add(converted.manufacturer_part_number);
                    parts.push(converted);
                }
            }
        }

        return jsonResponse(
            {
                query,
                success: true,
                error: null,
                parts,
                total_count: parts.length,
            },
            200,
            corsHeaders(origin),
        );
    } catch (err) {
        console.error("DigiKey search error:", err);
        return jsonResponse(
            {
                query,
                success: false,
                error:
                    err instanceof Error ? err.message : "Unknown error",
                parts: [],
                total_count: 0,
            },
            500,
            corsHeaders(origin),
        );
    }
}

// Main request handler
export default {
    async fetch(
        request: Request,
        env: Env,
        _ctx: ExecutionContext,
    ): Promise<Response> {
        const url = new URL(request.url);
        const path = url.pathname;
        const method = request.method;
        const origin = request.headers.get("Origin");

        // Handle CORS preflight
        if (method === "OPTIONS") {
            return new Response(null, {
                status: 204,
                headers: corsHeaders(origin),
            });
        }

        // Route requests
        try {
            // Auth routes
            if (path === "/auth/digikey/login" && method === "GET") {
                return handleLogin(request, env);
            }
            if (path === "/auth/digikey/callback" && method === "GET") {
                return handleCallback(request, env);
            }
            if (path === "/auth/digikey/logout" && method === "POST") {
                return handleLogout(request, env);
            }

            // API routes
            if (path === "/api/digikey/status" && method === "GET") {
                return handleStatus(request, env);
            }
            if (path === "/api/digikey/search" && method === "POST") {
                return handleSearch(request, env);
            }

            // Not found - pass through to static assets
            return new Response(null, { status: 404 });
        } catch (err) {
            console.error("Worker error:", err);
            return jsonResponse(
                {
                    error: "Internal server error",
                    message:
                        err instanceof Error ? err.message : "Unknown error",
                },
                500,
                corsHeaders(origin),
            );
        }
    },
};

